File: OsmAnd/src/net/osmand/plus/views/mapwidgets/widgets/NauticalGraphView.java
package net.osmand.plus.views.mapwidgets.widgets;

import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class NauticalGraphView extends View {
    private List<Double> data = new ArrayList<>();
    private final Paint linePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint textPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint gridPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private String unit = "";
    private final Path graphPath = new Path();
    private final Paint dotPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private float pulseRadius = 10f;
    private ValueAnimator pulseAnimator;

    public NauticalGraphView(Context context) {
        super(context);
        setWillNotDraw(false);
        init(context);
    }

    public NauticalGraphView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        setWillNotDraw(false);
        init(context);
    }

    public NauticalGraphView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setWillNotDraw(false);
        init(context);
    }

    private void init(Context context) {

        // Resolve theme colors manually to avoid 'styleable' and 'UiUtilities' errors
        TypedValue typedValue = new TypedValue();
        context.getTheme().resolveAttribute(android.R.attr.textColorPrimary, typedValue, true);
        int textColor = typedValue.data;

        context.getTheme().resolveAttribute(android.R.attr.colorForeground, typedValue, true);
        int gridColor = typedValue.data;

        // Grid Paint
        gridPaint.setColor(gridColor);
        gridPaint.setAlpha(100);
        gridPaint.setStyle(Paint.Style.STROKE);
        gridPaint.setStrokeWidth(2f);
        gridPaint.setAntiAlias(true);

        // Text Paint
        textPaint.setColor(textColor);
        textPaint.setTextSize(32f);
        textPaint.setAntiAlias(true);

        // Line Paint
        linePaint.setColor(Color.CYAN);
        linePaint.setStyle(Paint.Style.STROKE);
        linePaint.setStrokeWidth(8f);
        linePaint.setStrokeCap(Paint.Cap.ROUND);
        linePaint.setStrokeJoin(Paint.Join.ROUND);
        linePaint.setAntiAlias(true);

        // Dot Paint
        dotPaint.setColor(Color.WHITE);
        dotPaint.setStyle(Paint.Style.FILL);
        dotPaint.setAntiAlias(true);

        // Pulse animation
        pulseAnimator = ValueAnimator.ofFloat(10f, 20f);
        pulseAnimator.setDuration(1000);
        pulseAnimator.setRepeatCount(ValueAnimator.INFINITE);
        pulseAnimator.setRepeatMode(ValueAnimator.REVERSE);
        pulseAnimator.addUpdateListener(animation -> {
            pulseRadius = (float) animation.getAnimatedValue();
            postInvalidate();
        });
        pulseAnimator.start();
    }

    public void setData(List<Double> newData, String unit) {
        synchronized (this) {
            this.data = new ArrayList<>();
            for (Double val : newData) {
                if (val != null && val > 0 && val < 1000) {
                    this.data.add(val);
                }
            }
            this.unit = unit;
        }
        postInvalidate();
    }

    @Override
    protected void onDetachedFromWindow() {
        if (pulseAnimator != null) pulseAnimator.cancel();
        super.onDetachedFromWindow();
    }

    @Override
    protected void onDraw(@NonNull Canvas canvas) {
        super.onDraw(canvas);

        synchronized (this) {
            if (data == null || data.isEmpty()) {
                canvas.drawText("No Data", 20f, getHeight() / 2f, textPaint);
                return;
            }

            if (data.size() < 2) {
                canvas.drawText("Buffering...", 20f, getHeight() / 2f, textPaint);
                return;
            }

            float width = getWidth();
            float height = getHeight();
            float padding = 60f;
            float graphH = height - (padding * 2);

            // Find Range for scaling
            double min = Double.MAX_VALUE;
            double max = Double.MIN_VALUE;
            for (Double d : data) {
                if (d < min) min = d;
                if (d > max) max = d;
            }
            double range = (max - min) == 0 ? 1.0 : (max - min);
            double paddedMin = min - (range * 0.1);
            double paddedRange = range * 1.2;

            // 1. Draw Grid Lines
            canvas.drawLine(0, padding, width, padding, gridPaint); // Top
            canvas.drawLine(0, height - padding, width, height - padding, gridPaint); // Bottom

            // 2. Draw Data Path (Smoothed with Cubic Bezier if you want, or just lineTo)
            graphPath.reset();
            for (int i = 0; i < data.size(); i++) {
                float x = width - ((data.size() - 1 - i) * width) / (data.size() - 1);
                float y = (float) (height - padding - ((data.get(i) - paddedMin) / paddedRange) * graphH);

                if (i == 0) graphPath.moveTo(x, y);
                else graphPath.lineTo(x, y);
            }
            canvas.drawPath(graphPath, linePaint);

            // 3. Draw Axis Labels (AFTER drawing path so they are on top)
            // Y-Axis (Values) - Align Right, positioned just to the left of the padding line
            textPaint.setTextAlign(Paint.Align.RIGHT);
            canvas.drawText(String.format(Locale.US, "%.1f", max), padding - 5f, padding + 10f, textPaint);
            canvas.drawText(String.format(Locale.US, "%.1f", min), padding - 5f, height - padding + 5f, textPaint);

            // X-Axis (Time) - Align Center to the grid lines
            textPaint.setTextAlign(Paint.Align.CENTER);
            canvas.drawText("-6m", padding, height - 10f, textPaint);
            canvas.drawText("Now", width - padding, height - 10f, textPaint);

            // Unit Label - Top right
            textPaint.setTextAlign(Paint.Align.RIGHT);
            canvas.drawText(unit, width - 5f, padding - 10f, textPaint);

            // 4. Draw Current Value Indicator
            if (!data.isEmpty()) {
                // Always at the right edge
                float lastY = (float) (height - padding - ((data.get(data.size() - 1) - paddedMin) / paddedRange) * graphH);

                // Draw vertical "Now" marker
                gridPaint.setStrokeWidth(3f);
                canvas.drawLine(width, padding, width, height - padding, gridPaint);

                // Draw Pulsing Dot
                canvas.drawCircle(width, lastY, pulseRadius, dotPaint);

                // Draw text value next to the current point
                textPaint.setTextAlign(Paint.Align.RIGHT);
                canvas.drawText(String.format(Locale.US, "%.1f", data.get(data.size() - 1)), width - 15f, lastY - 15f, textPaint);
            }

        }
    }
}

File: OsmAnd/src/net/osmand/plus/views/mapwidgets/widgets/NauticalGraphView.kt
package net.osmand.plus.views.mapwidgets.widgets

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.util.TypedValue
import android.view.View
import androidx.core.graphics.toColorInt
import net.osmand.plus.R
import net.osmand.plus.OsmandApplication
import net.osmand.StateChangedListener
import net.osmand.plus.settings.enums.NauticalDisplayMode
import net.osmand.plus.plugins.nautical.ui.NauticalColorResolver
import net.osmand.plus.plugins.nautical.ui.NauticalSemanticColor
import java.util.*

class NauticalGraphView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
) : View(context, attrs, defStyleAttr) {

    private var data: MutableList<Pair<Double, Long>> = mutableListOf()
    private var dataCoeff = 1.0
    private var dataOffset = 0.0
    
    private val linePaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val textPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val gridPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val dotPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var unit = ""
    private val graphPath = Path()

    private var defaultLineColor = Color.CYAN
    private var defaultTextColor = Color.BLACK
    private var defaultGridColor = Color.LTGRAY

    init {
        setWillNotDraw(false)
        val density = resources.displayMetrics.density

        if (isInEditMode) {
            val now = System.currentTimeMillis()
            data.add(10.0 to (now - 3600000))
            data.add(15.0 to (now - 2700000))
            data.add(12.0 to (now - 1800000))
            data.add(20.0 to (now - 900000))
            data.add(18.0 to now)
            unit = "kn"
        }

        val typedValue = TypedValue()
        if (context.theme?.resolveAttribute(android.R.attr.textColorPrimary, typedValue, true) == true) {
            defaultTextColor = typedValue.data
        }

        if (context.theme?.resolveAttribute(android.R.attr.colorForeground, typedValue, true) == true) {
            defaultGridColor = typedValue.data
        }

        if (context.theme?.resolveAttribute(R.attr.active_color_primary, typedValue, true) == true) {
            defaultLineColor = typedValue.data
        }

        gridPaint.color = defaultGridColor
        gridPaint.alpha = 160
        gridPaint.style = Paint.Style.STROKE
        gridPaint.strokeWidth = density * 0.5f

        textPaint.color = defaultTextColor
        textPaint.textSize = 12f * density
        textPaint.typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)

        linePaint.color = defaultLineColor
        linePaint.style = Paint.Style.STROKE
        linePaint.strokeWidth = 2.5f * density
        linePaint.strokeCap = Paint.Cap.ROUND
        linePaint.strokeJoin = Paint.Join.ROUND

        dotPaint.color = defaultLineColor
        dotPaint.style = Paint.Style.FILL
    }

    fun setData(newData: List<Pair<Double, Long>>?, unit: String, coeff: Double = 1.0, offset: Double = 0.0) {
        if (newData == null) return
        synchronized(this) {
            data.clear()
            data.addAll(newData)
            this.unit = unit
            this.dataCoeff = coeff
            this.dataOffset = offset
        }
        postInvalidate()
    }

    private val nightVisionListener = StateChangedListener<NauticalDisplayMode> {
        postInvalidate()
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        val settings = (context.applicationContext as? OsmandApplication)?.settings
        settings?.NAUTICAL_DISPLAY_MODE?.addListener(nightVisionListener)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        val settings = (context.applicationContext as? OsmandApplication)?.settings
        settings?.NAUTICAL_DISPLAY_MODE?.removeListener(nightVisionListener)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        synchronized(this) {
            if (data.isEmpty()) {
                textPaint.textAlign = Paint.Align.CENTER
                textPaint.color = NauticalColorResolver.getColor(context, NauticalSemanticColor.SECONDARY)
                textPaint.alpha = 180
                textPaint.textSize = 20f * resources.displayMetrics.density
                canvas.drawText(context.getString(R.string.nautical_no_data), width / 2f, height / 2f, textPaint)
                return
            }

            val app = context.applicationContext as? OsmandApplication
            val mode = app?.settings?.NAUTICAL_DISPLAY_MODE?.get() ?: NauticalDisplayMode.NORMAL
            val isSunlight = mode == NauticalDisplayMode.SUNLIGHT
            val isNightVision = mode == NauticalDisplayMode.DARK

            linePaint.color = NauticalColorResolver.getColor(context, NauticalSemanticColor.PRIMARY)
            textPaint.color = NauticalColorResolver.getColor(context, NauticalSemanticColor.PRIMARY)
            gridPaint.color = NauticalColorResolver.getColor(context, NauticalSemanticColor.GRID)

            val width = width.toFloat()
            val height = height.toFloat()
            val density = resources.displayMetrics.density

            // Task 8.0: Extreme Display Adaptation
            if (isSunlight) {
                gridPaint.strokeWidth = density * 1.5f
                gridPaint.alpha = 255 // 100% alpha
                linePaint.strokeWidth = density * 4.0f
                linePaint.alpha = 255
            } else {
                gridPaint.strokeWidth = density * 0.5f
                gridPaint.alpha = 160
                linePaint.strokeWidth = density * 2.5f
            }

            if (isNightVision) {
                // Strict Scotopic Adaptation filter
                linePaint.color = 0xFFFF0000.toInt() // Pure Red
                textPaint.color = 0xFFFF0000.toInt()
                gridPaint.color = 0xFF330000.toInt() // Dark Red
                linePaint.alpha = 255
                textPaint.alpha = 255
                gridPaint.alpha = 255
            }

            val paddingH = 36f * density
            val paddingV = 16f * density
            val paddingBottom = 24f * density
            val graphW = width - (paddingH * 2)
            val graphH = height - (paddingV + paddingBottom)

            // Optimized min/max calculation and path building in one pass
            var min = Double.MAX_VALUE
            var max = -Double.MAX_VALUE
            
            for (p in data) {
                val v = p.first * dataCoeff + dataOffset
                if (v.isNaN() || v.isInfinite()) continue
                if (v < min) min = v
                if (v > max) max = v
            }

            if (min == Double.MAX_VALUE || max == -Double.MAX_VALUE) return

            if (min == max) {
                min -= 1.0
                max += 1.0
            }

            val range = max - min
            if (range <= 0 || range.isNaN() || range.isInfinite()) return
            val stepX = if (data.size > 1) graphW / (data.size - 1) else 0f
            if (stepX.isNaN() || stepX.isInfinite()) return

            // Draw Grid
            canvas.drawLine(paddingH, paddingV, width - paddingH, paddingV, gridPaint)
            canvas.drawLine(paddingH, height - paddingBottom, width - paddingH, height - paddingBottom, gridPaint)

            // Draw Path
            graphPath.reset()
            for (i in data.indices) {
                val x = paddingH + (i * stepX)
                val v = data[i].first * dataCoeff + dataOffset
                val y = (height - paddingBottom - (((v - min) / range) * graphH)).toFloat()
                if (i == 0) graphPath.moveTo(x, y)
                else graphPath.lineTo(x, y)
            }
            canvas.drawPath(graphPath, linePaint)

            // Draw Labels
            textPaint.textAlign = Paint.Align.RIGHT
            val shadowColor = if (!isNightVision && Color.luminance(textPaint.color) > 0.5f) Color.BLACK else Color.WHITE
            if (!isNightVision) textPaint.setShadowLayer(2f, 1f, 1f, shadowColor)

            canvas.drawText(String.format(Locale.US, "%.1f", max), paddingH - (4 * density), paddingV + (4 * density), textPaint)
            canvas.drawText(String.format(Locale.US, "%.1f", min), paddingH - (4 * density), height - paddingBottom, textPaint)

            textPaint.textAlign = Paint.Align.LEFT
            canvas.drawText(unit, (width - paddingH) + (4 * density), paddingV + (4 * density), textPaint)
            
            val now = System.currentTimeMillis()
            val totalSpanMs = now - data.first().second
            if (totalSpanMs > 1000) {
                textPaint.textAlign = Paint.Align.CENTER
                val spanMinutes = totalSpanMs / 60000
                if (spanMinutes > 0) {
                    canvas.drawText("-${spanMinutes}m", paddingH, height - (4 * density), textPaint)
                    canvas.drawText("-${spanMinutes / 2}m", paddingH + (graphW / 2), height - (4 * density), textPaint)
                } else {
                    canvas.drawText("-${totalSpanMs / 1000}s", paddingH, height - (4 * density), textPaint)
                }
            }
            textPaint.clearShadowLayer()
        }
    }
}
